File: app/src/main/java/life/plank/juna/zone/view/activity/CreateBoardActivity.java
package life.plank.juna.zone.view.activity;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.provider.MediaStore.Images.Media;
import android.support.design.widget.TextInputEditText;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Button;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ToggleButton;

import java.util.Objects;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import life.plank.juna.zone.R;
import life.plank.juna.zone.ZoneApplication;
import life.plank.juna.zone.data.model.Board;
import life.plank.juna.zone.util.UIDisplayUtil;
import life.plank.juna.zone.util.customview.ZoneToolBar;
import life.plank.juna.zone.view.adapter.BoardColorThemeAdapter;
import life.plank.juna.zone.view.adapter.BoardIconAdapter;

import static com.facebook.internal.Utility.isNullOrEmpty;
import static life.plank.juna.zone.util.AppConstants.GALLERY_IMAGE_RESULT;
import static life.plank.juna.zone.util.UIDisplayUtil.enableOrDisableView;
import static life.plank.juna.zone.util.UIDisplayUtil.getPathForGalleryImageView;
import static life.plank.juna.zone.util.UIDisplayUtil.loadBitmap;
import static life.plank.juna.zone.util.UIDisplayUtil.toggleZone;

public class CreateBoardActivity extends AppCompatActivity {

    public static Bitmap parentViewBitmap = null;

    @BindView(R.id.parent_layout)
    ScrollView parentLayout;
    @BindView(R.id.football)
    ToggleButton football;
    @BindView(R.id.music)
    ToggleButton music;
    @BindView(R.id.drama)
    ToggleButton drama;
    @BindView(R.id.tune)
    ToggleButton tune;
    @BindView(R.id.skill)
    ToggleButton skill;
    @BindView(R.id.other)
    ToggleButton other;
    @BindView(R.id.board_name_edit_text)
    TextInputEditText boardName;
    @BindView(R.id.board_description_edit_text)
    TextInputEditText boardDescription;
    @BindView(R.id.private_board_color_list)
    RecyclerView privateBoardColorList;
    @BindView(R.id.private_board_icon_list)
    RecyclerView privateBoardIconList;
    @BindView(R.id.upload_board_icon)
    Button uploadBoardIcon;
    @BindView(R.id.create_board_button)
    Button createPrivateBoard;
    @BindView(R.id.tool_bar)
    ZoneToolBar toolbar;
    @BindView(R.id.user_greeting)
    TextView userGreeting;

    @Inject
    BoardColorThemeAdapter boardColorThemeAdapter;
    @Inject
    BoardIconAdapter boardIconAdapter;

    private ToggleButton[] zones;
    private String zone;
    private String filePath;
    private Boolean isIconSelected = false;

    public static void launch(Context packageContext, String username) {
        Intent intent = new Intent(packageContext, CreateBoardActivity.class);
        intent.putExtra(ZoneApplication.getContext().getString(R.string.username), username);
        packageContext.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_board);
        ButterKnife.bind(this);

        ((ZoneApplication) getApplication()).getUiComponent().inject(this);

        zones = new ToggleButton[]{football, music, drama, tune, skill, other};
        privateBoardColorList.setAdapter(boardColorThemeAdapter);
        privateBoardIconList.setAdapter(boardIconAdapter);

        UIDisplayUtil.checkPermission(CreateBoardActivity.this);
        boardIconAdapter.boardIconList.clear();

        validateCreateBoardContent();

        SharedPreferences sharedPref = getSharedPreferences(getString(R.string.pref_user_details), MODE_PRIVATE);
        if (!sharedPref.getString(getString(R.string.pref_profile_pic_url), getString(R.string.na)).equals(getString(R.string.na))) {
            toolbar.setProfilePic(sharedPref.getString(getString(R.string.pref_profile_pic_url), getString(R.string.na)));
        }
        userGreeting.setText(getString(R.string.hi_user, getIntent().getStringExtra(ZoneApplication.getContext().getString(R.string.username))));
    }

    @OnTextChanged({R.id.board_name_edit_text, R.id.board_description_edit_text})
    public void onTextFieldUpdated() {
        validateCreateBoardContent();
    }

    @OnClick({R.id.football, R.id.music, R.id.drama, R.id.tune, R.id.skill, R.id.other})
    public void toggleView(ToggleButton view) {
        for (ToggleButton zoneView : zones) {
            if (zoneView.getId() == view.getId()) {
                checkAction(zoneView, !zoneView.isChecked());
                continue;
            }
            checkAction(zoneView, false);
        }
        validateCreateBoardContent();
    }

    private void checkAction(ToggleButton toggleButton, boolean isChecked) {
        if (Objects.equals(zone, toggleButton.getText().toString())) {
            if (!isChecked) {
                zone = null;
            }
        } else {
            if (isChecked) {
                zone = toggleButton.getText().toString();
            }
        }
        toggleZone(this, toggleButton, isChecked);
    }

    @OnClick(R.id.create_board_button)
    public void onViewClicked(View view) {
        Board board = new Board(
                boardName.getText().toString().trim(),
                getString(R.string.private_lowercase),
                zone.toLowerCase().trim(),
                boardDescription.getText().toString().trim(),
                boardColorThemeAdapter.getSelectedColor()
        );

        createBoard(board, boardIconAdapter.getSelectedPath());
    }

    @OnClick(R.id.upload_board_icon)
    public void onButtonClicked(View view) {
        if (UIDisplayUtil.checkPermission(CreateBoardActivity.this)) {
            getImageResourceFromGallery();
        } else {
            Toast.makeText(this, R.string.add_permission, Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case GALLERY_IMAGE_RESULT:
                switch (resultCode) {
                    case RESULT_OK:
                        filePath = getPathForGalleryImageView(data.getData(), this);
                        if (filePath != null) {
                            boardIconAdapter.boardIconList.add(0, filePath);
                            boardIconAdapter.setSelectedIndex(0);
                            boardIconAdapter.notifyItemInserted(0);
                            isIconSelected = true;
                            validateCreateBoardContent();
                        } else {
                            Toast.makeText(this, R.string.image_not_supported, Toast.LENGTH_SHORT).show();
                        }
                        break;

                    case RESULT_CANCELED:
                        finish();
                        break;
                    default:
                        Toast.makeText(this, R.string.failed_to_process_image, Toast.LENGTH_LONG).show();
                        finish();
                        break;
                }
                break;
            default:
                break;
        }
    }

    public void getImageResourceFromGallery() {
        Intent galleryIntent = new Intent(Intent.ACTION_PICK);
        galleryIntent.setDataAndType(Media.EXTERNAL_CONTENT_URI, getString(R.string.image_format));
        startActivityForResult(galleryIntent, GALLERY_IMAGE_RESULT);
    }

    private void validateCreateBoardContent() {
        enableOrDisableView(
                createPrivateBoard,
                !(isNullOrEmpty(zone) || isNullOrEmpty(zone.toLowerCase().trim()) ||
                        boardName.getText() == null || isNullOrEmpty(boardName.getText().toString().trim()) ||
                        boardDescription.getText() == null || isNullOrEmpty(boardDescription.getText().toString().trim()) ||
                        isNullOrEmpty(boardColorThemeAdapter.getSelectedColor()) || !isIconSelected)
        );
    }

    private void createBoard(Board board, String file) {
        if (isNullOrEmpty(file)) {
            Toast.makeText(this, R.string.select_image_to_upload, Toast.LENGTH_SHORT).show();
            return;
        }

        parentViewBitmap = loadBitmap(getWindow().getDecorView(), getWindow().getDecorView(), this);
        BoardPreviewActivity.launch(this, board, file);
    }
}

File: app/src/main/java/life/plank/juna/zone/view/activity/board/CreateBoardActivity.kt
package life.plank.juna.zone.view.activity.board

import android.app.Activity
import android.content.*
import android.os.Bundle
import android.provider.MediaStore.Images.Media
import android.view.View
import android.widget.ToggleButton
import com.prembros.facilis.util.*
import kotlinx.android.synthetic.main.activity_create_board.*
import life.plank.juna.zone.*
import life.plank.juna.zone.data.model.Board
import life.plank.juna.zone.data.network.interfaces.RestApi
import life.plank.juna.zone.util.common.AppConstants.GALLERY_IMAGE_RESULT
import life.plank.juna.zone.util.common.customToast
import life.plank.juna.zone.util.sharedpreference.PreferenceManager
import life.plank.juna.zone.util.view.UIDisplayUtil
import life.plank.juna.zone.util.view.UIDisplayUtil.*
import life.plank.juna.zone.view.activity.base.BaseJunaCardActivity
import life.plank.juna.zone.view.adapter.board.creation.*
import life.plank.juna.zone.view.fragment.board.user.BoardPreviewPopup
import org.jetbrains.anko.sdk27.coroutines.onClick
import javax.inject.Inject

class CreateBoardActivity : BaseJunaCardActivity() {

    @Inject
    lateinit var boardColorThemeAdapter: BoardColorThemeAdapter
    @Inject
    lateinit var boardIconAdapter: BoardIconAdapter

    private var zones: Array<ToggleButton>? = null
    private var zone: String? = null
    private var filePath: String? = null
    private var isIconSelected: Boolean? = false

    companion object {
        fun launch(packageContext: Context, username: String) {
            val intent = Intent(packageContext, CreateBoardActivity::class.java)
            intent.putExtra(ZoneApplication.getContext().getString(R.string.username), username)
            packageContext.startActivity(intent)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_board)
        ZoneApplication.getApplication().uiComponent.inject(this)

        zones = arrayOf(football, music, drama, tune, skill, other)
        private_board_color_list.adapter = boardColorThemeAdapter
        private_board_icon_list.adapter = boardIconAdapter

        UIDisplayUtil.checkPermission(this@CreateBoardActivity)
        boardIconAdapter.boardIconList.clear()

        validateCreateBoardContent()

        PreferenceManager.CurrentUser.getProfilePicUrl()?.run { tool_bar.setProfilePic(this) }
        tool_bar.isNotificationViewVisible(View.GONE)
        user_greeting.text = getString(R.string.hi_user, intent.getStringExtra(ZoneApplication.getContext().getString(R.string.username)))

        setupListeners()
    }

    override fun getFragmentContainer(): Int = R.id.board_maker_fragment_container

    override fun restApi(): RestApi? = null

    private fun setupListeners() {
        arrayOf(football, music, drama, tune, skill, other).onClick { toggleView(it as ToggleButton) }

        arrayOf(board_name_edit_text, board_description_edit_text).onTextChanged { validateCreateBoardContent() }

        create_board_button.onClick { onCreateBoardButtonClicked() }

        upload_board_icon.onClick { onUploadButtonClicked() }
    }

    private fun toggleView(view: ToggleButton) {
        for (zoneView in zones!!) {
            if (zoneView.id == view.id) {
                checkAction(zoneView, !zoneView.isChecked)
                continue
            }
            checkAction(zoneView, false)
        }
        validateCreateBoardContent()
    }

    private fun checkAction(toggleButton: ToggleButton, isChecked: Boolean) {
        if (zone == toggleButton.text.toString()) {
            if (!isChecked) {
                zone = null
            }
        } else {
            if (isChecked) {
                zone = toggleButton.text.toString()
            }
        }
        toggleZone(this, toggleButton, isChecked)
    }

    private fun onCreateBoardButtonClicked() {
        board_name_edit_text.clearFocus()
        board_description_edit_text.clearFocus()
        hideSoftKeyboard(create_board_button)

        val board = Board(
                board_name_edit_text.text!!.toString().trim(),
                getString(R.string.private_lowercase),
                zone!!.toLowerCase().trim(),
                board_description_edit_text.text!!.toString().trim(),
                boardColorThemeAdapter.selectedColor!!
        )

        boardIconAdapter.selectedPath?.run { createBoard(board, this) }
    }

    private fun onUploadButtonClicked() {
        if (UIDisplayUtil.checkPermission(this@CreateBoardActivity)) {
            getImageResourceFromGallery()
        } else {
            customToast(R.string.add_permission)
        }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            GALLERY_IMAGE_RESULT -> when (resultCode) {
                Activity.RESULT_OK -> {
                    filePath = getPathForGalleryImageView(data!!.data, this)
                    if (filePath != null) {
                        boardIconAdapter.boardIconList.add(0, filePath!!)
                        boardIconAdapter.selectedIndex = 0
                        boardIconAdapter.selectedPath = filePath
                        boardIconAdapter.notifyDataSetChanged()
                        isIconSelected = true
                        validateCreateBoardContent()
                    } else {
                        customToast(R.string.image_not_supported)
                    }
                }
                Activity.RESULT_CANCELED -> {/*Do nothing*/
                }
                else -> customToast(R.string.failed_to_process_image)
            }
        }
    }

    private fun getImageResourceFromGallery() {
        val galleryIntent = Intent(Intent.ACTION_PICK)
        galleryIntent.setDataAndType(Media.EXTERNAL_CONTENT_URI, getString(R.string.image_format))
        startActivityForResult(galleryIntent, GALLERY_IMAGE_RESULT)
    }

    private fun validateCreateBoardContent() {
        enableOrDisableView(
                create_board_button,
                !(isNullOrEmpty(zone) || isNullOrEmpty(zone!!.toLowerCase().trim()) ||
                        board_name_edit_text.text == null || isNullOrEmpty(board_name_edit_text.text!!.toString().trim()) ||
                        board_description_edit_text.text == null || isNullOrEmpty(board_description_edit_text.text!!.toString().trim()) ||
                        isNullOrEmpty(boardColorThemeAdapter.selectedColor) || (!isIconSelected!!))
        )
    }

    private fun createBoard(board: Board, file: String) {
        if (isNullOrEmpty(file)) {
            customToast(R.string.select_image_to_upload)
            return
        }
        pushPopup(BoardPreviewPopup.newInstance(board, file))
    }

    override fun onBackPressed() {
        if (supportFragmentManager.removeActivePopupsIfAny()) {
            finish()
        }
    }
}